internal/checker/checker.go

```diff
@@ -2306,6 +2310,60 @@ func (c *Checker) checkSourceElementWorker(node *ast.Node) {
 	}
 }
 
+func (c *Checker) checkSourceElementUnreachable(node *ast.Node) bool {
+	if c.reportedUnreachableNodes.Has(node) {
+		return true
+	}
+
+	if !c.isSourceElementUnreachable(node) {
+		return false
+	}
+
+	c.reportedUnreachableNodes.Add(node)
+
+	isError := c.compilerOptions.AllowUnreachableCode == core.TSFalse
+	sourceFile := ast.GetSourceFileOfNode(node)
+
+	start := scanner.GetRangeOfTokenAtPosition(sourceFile, node.Pos()).Pos()
+	end := node.End()
+
+	parent := node.Parent
+	if parent.CanHaveStatements() {
+		statements := parent.Statements()
+		if offset := slices.Index(statements, node); offset >= 0 {
+			for _, nextNode := range statements[offset+1:] {
+				if !c.isSourceElementUnreachable(nextNode) {
+					break
+				}
+				end = nextNode.End()
+				c.reportedUnreachableNodes.Add(nextNode)
+			}
+		}
+	}
+
+	diagnostic := ast.NewDiagnostic(sourceFile, core.NewTextRange(start, end), diagnostics.Unreachable_code_detected)
+	c.addErrorOrSuggestion(isError, diagnostic)
+
+	return true
+}
+
+func (c *Checker) isSourceElementUnreachable(node *ast.Node) bool {
+	if node.Flags&ast.NodeFlagsUnreachable != 0 {
+		switch node.Kind {
+		case ast.KindEnumDeclaration:
+			return !ast.IsEnumConst(node) || c.compilerOptions.ShouldPreserveConstEnums()
+		case ast.KindModuleDeclaration:
+			return ast.IsInstantiatedModule(node, c.compilerOptions.ShouldPreserveConstEnums())
+		default:
+			return true
+		}
+	} else if ast.KindFirstStatement <= node.Kind && node.Kind <= ast.KindLastStatement {
```

Comment:
Can you leave a comment either here or at `NodeFlagsUnreachable` about why only one of these checks isn't sufficient?